src/environment.js

```diff
@@ -53,34 +124,66 @@ class Environment {
         return commitSha !== 'HEAD' ? this._env.BUILDKITE_COMMIT : null;
       }
     }
+
     return null;
   }
 
   get branch() {
     if (this._env.PERCY_BRANCH) {
       return this._env.PERCY_BRANCH;
     }
+    let result = '';
     switch (this.ci) {
       case 'travis':
         if (this.pullRequestNumber && this._env.TRAVIS_PULL_REQUEST_BRANCH) {
-          return this._env.TRAVIS_PULL_REQUEST_BRANCH;
+          result = this._env.TRAVIS_PULL_REQUEST_BRANCH;
+        } else {
+          result = this._env.TRAVIS_BRANCH;
         }
-        return this._env.TRAVIS_BRANCH;
+        break;
       case 'jenkins':
-        return this._env.ghprbSourceBranch;
+        result = this._env.ghprbSourceBranch;
+        break;
       case 'circle':
-        return this._env.CIRCLE_BRANCH;
+        result = this._env.CIRCLE_BRANCH;
+        break;
       case 'codeship':
-        return this._env.CI_BRANCH;
+        result = this._env.CI_BRANCH;
+        break;
       case 'drone':
-        return this._env.DRONE_BRANCH;
+        result = this._env.DRONE_BRANCH;
+        break;
       case 'semaphore':
-        return this._env.BRANCH_NAME;
+        result = this._env.BRANCH_NAME;
+        break;
       case 'buildkite':
-        return this._env.BUILDKITE_BRANCH;
+        result = this._env.BUILDKITE_BRANCH;
+        break;
+    }
+
+    if (result == '') {
+      result = this.rawBranch();
+    }
+
+    if (result == '') {
+      // Branch not specified
+      result = null;
+    }
+
+    return result;
+  }
+
+  rawBranch() {
+    const child_process = require('child_process');
```

Comment:
I notice this same sort of setup twice. Optionally consider creating a method like `execGit`.

Something like this:

```javascript
execGit(command) {
  try {
    return child_process
      .execSync(`git ${cmd}`)
      .toString()
      .trim();
  } catch (error) {
    return '';
  }
}
```

Then you could have

```javascript
execGit('rev-parse --abbrev-ref HEAD')
```

Reply:
Good call!  Will do.